components/builder-depot/src/server.rs

```diff
@@ -439,55 +446,91 @@ fn write_archive(filename: &PathBuf, body: &mut Body) -> Result<PackageArchive>
+                    Err(Error::NetError(e)) => return Ok(render_net_error(&e)),
+                    Err(_) => unreachable!(),
+                }
+            } else {
+                return Ok(render_net_error(&err));
+            }
+        }
     };
 
     let xfilename = format!("{}-{}.pub", key.get_name(), key.get_revision());
     download_content_as_file(key.get_body(), xfilename)
 }
 
-fn generate_origin_encryption_keys(req: &mut Request) -> IronResult<Response> {
-    debug!("Generate Origin Encryption Keys {:?}", req);
-    match get_param(req, "origin") {
-        Some(origin) => {
-            if !check_origin_access(req, &origin).unwrap_or(false) {
-                return Ok(Response::with(status::Forbidden));
-            }
+fn generate_origin_encryption_keys(
+    origin: &Origin,
+    req: &mut Request,
+) -> Result<OriginPublicEncryptionKey> {
+    debug!("Generate Origin Encryption Keys {:?} for {:?}", req, origin);
 
-            match helpers::get_origin(req, origin) {
-                Ok(origin) => {
-                    match helpers::generate_origin_encryption_keys(req, origin) {
-                        Ok(_) => Ok(Response::with(status::Created)),
-                        Err(err) => Ok(render_net_error(&err)),
-                    }
-                }
-                Err(err) => Ok(render_net_error(&err)),
-            }
-        }
-        None => Ok(Response::with(status::BadRequest)),
-    }
+    let mut public_request = OriginPublicEncryptionKeyCreate::new();
+    let mut private_request = OriginPrivateEncryptionKeyCreate::new();
+    let mut public_key = OriginPublicEncryptionKey::new();
+    let mut private_key = OriginPrivateEncryptionKey::new();
+    {
+        let session = req.extensions.get::<Authenticated>().unwrap();
+        public_key.set_owner_id(session.get_id());
+        private_key.set_owner_id(session.get_id());
+    }
+    public_key.set_name(origin.get_name().to_string());
+    public_key.set_origin_id(origin.get_id());
+    private_key.set_name(origin.get_name().to_string());
+    private_key.set_origin_id(origin.get_id());
+
+    let pair = BoxKeyPair::generate_pair_for_origin(origin.get_name())
+        .map_err(Error::HabitatCore)?;
+    public_key.set_revision(pair.rev.clone());
+    public_key.set_body(
+        pair.to_public_string()
+            .map_err(Error::HabitatCore)?
+            .into_bytes(),
+    );
+    private_key.set_revision(pair.rev.clone());
+    private_key.set_body(
+        pair.to_secret_string()
+            .map_err(Error::HabitatCore)?
+            .into_bytes(),
+    );
+
+    public_request.set_public_encryption_key(public_key);
+    private_request.set_private_encryption_key(private_key);
+
+    let key = route_message::<OriginPublicEncryptionKeyCreate, OriginPublicEncryptionKey>(
+        req,
+        &public_request,
+    )?;
+    route_message::<OriginPrivateEncryptionKeyCreate, OriginPrivateEncryptionKey>(
+        req,
+        &private_request,
+    )?;
```

Comment:
So if I'm understanding the existing code correctly, if someone hits this function and the call to create the public key succeeds, and the call to create the private key fails for some transient reason, it's not a big deal.  We'll return an error to the user, but if the user tries again (assuming the transient issue has resolved), it will work and they'll be all set.  From what I could tell, the only unique constraint we have is on the full name of the key, which includes the revision, and the revision will change every time they generate a new key.  Do I have all that correct?

If so, then this seems fine as is.

Reply:
yep that is all correct. I am thinking about revisiting the logic there to do an update on the public keys but that's a future iteration